src/core/lib/iomgr/combiner.cc

```diff
@@ -343,6 +345,39 @@ static void combiner_finally_exec(grpc_closure* closure, grpc_error* error) {
   grpc_closure_list_append(&lock->final_list, closure, error);
 }
 
+static void combiner_run(grpc_closure* closure, grpc_error* error) {
+#ifndef NDEBUG
+  closure->scheduled = false;
+  grpc_combiner* lock = COMBINER_FROM_CLOSURE_SCHEDULER(closure, scheduler);
+  GRPC_COMBINER_TRACE(gpr_log(
+      GPR_DEBUG,
+      "Combiner:%p grpc_combiner_run closure:%p created [%s:%d] run [%s:%d]",
+      lock, closure, closure->file_created, closure->line_created,
+      closure->file_initiated, closure->line_initiated));
+  GPR_ASSERT(grpc_core::ExecCtx::Get()->combiner_data()->active_combiner ==
```

Comment:
grpc_core::ExecCtx::Get() is a pointer thread local so there are no atomics there. As for being blind to bugs, I was hoping that all such scheduling issues (which are going to be programmer errors) would be caught in our tests in debug mode.

Also, it won't be correct to run a closure of a different combiner/scheduler while running inside a combiner. That would lead to all kinds of synchronization issues. A better option would be to just use closure_sched in that case. This also means that we can't just blindly convert closure_sched to closure_run everywhere.

Reply:
I could also let combiner_run fallback on combiner_exec if the condition is not satisfied instead of asserting out.

Reply:
I think that as long as we have separate `GRPC_CLOSURE_RUN()` and `GRPC_CLOSURE_SCHED()` calls and require callers to choose the right one, then we should make this an assertion to ensure that all callers are doing the right thing.  However, if we ever decided to switch to a model where we combine the two calls and expect the implementation to do the right thing based on the scheduler, then we could have the implementation switch dynamically.